packager/VisibilityChecker.cc

```diff
@@ -0,0 +1,558 @@
+
+        if (sym.exists()) {
+            sym.data(gs)->flags.isExported = true;
+        }
+    }
+
+    // Mark both the package and its test namespace as exported. This will terminate the recursive exporting in
+    // `exportParentNamespace`, as it stops when it hits either the root or an exported symbol.
+    void exportPackageRoots(core::GlobalState &gs) {
+        this->exportRoot(gs, core::Symbols::root());
+
+        auto test = core::Symbols::root().data(gs)->findMember(gs, core::Names::Constants::Test());
+        if (test.exists() && test.isClassOrModule()) {
+            this->exportRoot(gs, test.asClassOrModuleRef());
+        }
+    }
+
+    void setPackageLocs(core::MutableContext ctx, core::LocOffsets loc, core::ClassOrModuleRef sym) {
+        std::vector<core::NameRef> names;
+
+        while (sym.exists() && sym != core::Symbols::PackageSpecRegistry()) {
+            // The symbol isn't a package name if it's defined outside of the package registry.
+            if (sym == core::Symbols::root()) {
+                return;
+            }
+
+            names.emplace_back(sym.data(ctx)->name);
+            sym = sym.data(ctx)->owner;
+        }
+
+        absl::c_reverse(names);
+
+        auto &db = ctx.state.packageDB();
+
+        {
+            auto packageSym = core::Symbols::root();
+            for (auto name : names) {
+                auto member = packageSym.data(ctx)->findMember(ctx, name);
+                if (!member.exists() || !member.isClassOrModule()) {
+                    packageSym = core::Symbols::noClassOrModule();
+                    break;
+                }
+
+                packageSym = member.asClassOrModuleRef();
+            }
+
+            if (packageSym.exists()) {
+                auto file = packageSym.data(ctx)->loc().file();
+                if (db.getPackageNameForFile(file) != this->package.mangledName()) {
+                    packageSym.data(ctx)->addLoc(ctx, ctx.locAt(loc));
+                }
+            }
+        }
+
+        {
+            auto testSym = core::Symbols::root();
+            auto member = testSym.data(ctx)->findMember(ctx, core::Names::Constants::Test());
+            if (!member.exists() || !member.isClassOrModule()) {
+                return;
+            }
+
+            testSym = member.asClassOrModuleRef();
+            for (auto name : names) {
+                auto member = testSym.data(ctx)->findMember(ctx, name);
+                if (!member.exists() || !member.isClassOrModule()) {
+                    testSym = core::Symbols::noClassOrModule();
+                    break;
+                }
+
+                testSym = member.asClassOrModuleRef();
+            }
+
+            if (testSym.exists()) {
+                auto file = testSym.data(ctx)->loc().file();
+                if (db.getPackageNameForFile(file) != this->package.mangledName()) {
+                    testSym.data(ctx)->addLoc(ctx, ctx.locAt(loc));
+                }
+            }
```

Comment:
So this would mean that

```ruby
class A::B < PackageSpec
end
```

makes it so that

```ruby
class Test::A::B
  #            ^ jump-to-def
end
```

makes the jump to def go to the package spec file (among potentially other files)? is that what we desire and/or for compat with existing behavior?

Reply:
I don't believe that will happen in this case: that would only happen in the case that the package name is never defined explicitly as a class or module. A case that looks like the following:
```ruby
class A::B < PackageSpec
end

class Test::A::B::C
  #            ^ jump-to-def
end
```

is what would trigger the use of `addLoc`, to ensure that we have an authoritative location for `Test::A::B` that marks it as part of the `A::B` package. Without this, if there's a sub-package definition of `A::B::OtherPackage`, and someone refers to `Test::A::B`, it's a toss-up as to whether the owning package of `Test::A::B` will be `A::B` or `A::B::OtherPackage`.

Reply:
That makes sense, thanks!